torchmetrics/functional/classification/confusion_matrix.py

```diff
@@ -33,23 +33,30 @@ def _confusion_matrix_update(
             case of binary or multi-label inputs. Default value of 0.5 corresponds to input being probabilities.
         multilabel: determines if data is multilabel or not.
     """
-
-    preds, target, mode = _input_format_classification(preds, target, threshold)
-    if mode not in (DataType.BINARY, DataType.MULTILABEL):
-        preds = preds.argmax(dim=1)
-        target = target.argmax(dim=1)
-    if multilabel:
-        unique_mapping = ((2 * target + preds) + 4 * torch.arange(num_classes, device=preds.device)).flatten()
-        minlength = 4 * num_classes
-    else:
-        unique_mapping = (target.view(-1) * num_classes + preds.view(-1)).to(torch.long)
-        minlength = num_classes ** 2
-
-    bins = torch.bincount(unique_mapping, minlength=minlength)
-    if multilabel:
-        confmat = bins.reshape(num_classes, 2, 2)
-    else:
-        confmat = bins.reshape(num_classes, num_classes)
+    input_threshold, input_preds, input_target = threshold, preds, target
+    thresholds = threshold if isinstance(threshold, (list, Tensor)) else [threshold]
+
+    confmat_list = []
+    for threshold in thresholds:
+        preds, target, mode = _input_format_classification(input_preds, input_target, threshold)
+        if mode not in (DataType.BINARY, DataType.MULTILABEL):
+            preds = preds.argmax(dim=1)
+            target = target.argmax(dim=1)
+        if multilabel:
+            unique_mapping = ((2 * target + preds) + 4 * torch.arange(num_classes, device=preds.device)).flatten()
+            minlength = 4 * num_classes
+        else:
+            unique_mapping = (target.view(-1) * num_classes + preds.view(-1)).to(torch.long)
+            minlength = num_classes ** 2
+        bins = torch.bincount(unique_mapping, minlength=minlength)
+        if multilabel:
+            confmat_list.append(bins.reshape(num_classes, 2, 2))
+        else:
+            confmat_list.append(bins.reshape(num_classes, num_classes))
+    confmat = torch.stack(confmat_list)
+    # If a single threshold is provided, return the confusion matrix without the thresholds dimension
+    if isinstance(input_threshold, float):
+        confmat = confmat.squeeze(dim=0)
```

Comment:
I think any reasonably efficient implementation of these metrics needs for this code to be vectorised (presumably like we have done it in [seer-algo](https://github.com/seermedical/seer-algo/blob/dcf53db78117401835e7e856ed4ae54b62f5c6db/seeralgo/metrics/metrics.py#L136). The original approach used doesn't lend itself well to doing that since it relies on first modifying the `preds` before using `argmax` (and to vectorise that we'd need to create a huge number of copies of `preds`).

I think we might be best off just implementing a new `ConfusionMatrixCurve` or similar and implementing in the optimal way. Then we can find other places in the existing binned metrics that can take advantage of it